src/Bicep.Core/SourceCode/SourceArchive.cs

```diff
@@ -68,9 +89,32 @@ private partial record SourceFileInfoEntry(
             string Kind         // kind of source
         );
 
-        public static SourceArchive FromStream(Stream stream)
+        private string? GetRequiredBicepVersionMessage()
         {
-            return new SourceArchive(stream);
+            if (MetadataVersion < CurrentMetadataVersion)
+            {
+                return $"This source code was published with an older, incompatible version of Bicep ({BicepVersion}). You are using version {ThisAssembly.AssemblyVersion}.";
+            }
+
+            if (MetadataVersion > CurrentMetadataVersion)
+            {
+                return $"This source code was published with a newer, incompatible version of Bicep ({BicepVersion}). You are using version {ThisAssembly.AssemblyVersion}. You need a newer version in order to view the module source.";
+            }
+
+            return null;
+        }
+
+        public static ResultWithException<SourceArchive> UnpackFromStream(Stream stream)
```

Comment:
It may be better to operate on `BinaryData` than `Stream` since a stream may have been unintentionally disposed of or closed. Also the need to keep reseting the Stream position back to 0 feels wrong.

Reply:
Removed one stream reset that was definitely incorrect (5ba82f903).  As for the one at the end of PackSourcesIntoStream, that's simply necessary since I'm returning a new stream that I just wrote.  

I tend to agree binary data would be easier to deal with, but all the surrounding code for publishing artifacts uses streams.  Maybe we should change it sometime.  Think it's out of scope here.